activesupport/lib/active_support/messages/metadata.rb

```diff
@@ -82,7 +82,7 @@ def extract_from_metadata_envelope(envelope, purpose: nil)
             throw :invalid_message_content, "expired"
           end
 
-          if hash["pur"] != purpose&.to_s
+          if hash["pur"].to_s != purpose.to_s
```

Comment:
Wouldn't this `to_s` result in matching purpose for `"1"` and `1`?

Reply:
Correct, the new serialization [already transforms the purpose in a String](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/messages/metadata.rb#L67) so `hash["pur"]` is always a String if serialized with the new logic.
[The old serialization, on the other hand doesn't do it](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/messages/metadata.rb#L74) and this introduces the mismatch triggering the issue. 
Isn't fine to always consider purpose as a String? We are already doing it after all.

Reply:
It seems like the old comparison version was also calling `to_s`:
https://github.com/rails/rails/blame/ebc3b660e51c5bdfd639639bb3c605d522d1a728/activesupport/lib/active_support/messages/metadata.rb#L67